src/logger/BaseLogger.go

```diff
@@ -0,0 +1,144 @@
+// import (
+// 	"net/url"
+// )
+
+/**
+ * Initialize enabled logger using default url.
+ */
+func NewBaseLoggerAgent(_agent string) *BaseLogger {
+	return NewBaseLogger(_agent, "insert default url here", true, nil)
+}
+
+/**
+ * Initialize enabled/disabled logger using default url.
+ */
+func NewBaseLoggerAgentEnabled(_agent string, _enabled bool) *BaseLogger {
+	return NewBaseLogger(_agent, "insert default url here", _enabled, nil)
+}
+
+/**
+ * Initialize enabled logger using url.
+ */
+func NewBaseLoggerAgentUrl(_agent string, _url string) *BaseLogger {
+	return NewBaseLogger(_agent, _url, true, nil)
+}
+
+/**
+ * Initialize enabled logger using queue.
+ */
+func NewBaseLoggerAgentQueue(_agent string, _queue []string) *BaseLogger {
+	return NewBaseLogger(_agent, "insert default url here", true, nil)
+}
+
+/**
+* Initialize enabled/disabled logger using queue.
+ */
+func NewBaseLoggerAgentQueueEnabled(_agent string, _queue []string, _enabled bool) *BaseLogger {
+	return NewBaseLogger(_agent, "insert default url here", _enabled, nil)
+}
+
+//main constructor
+func NewBaseLogger(_agent string, _url string, _enabled bool, _queue []string) *BaseLogger {
+	baselogger := &BaseLogger{}
+	baselogger.agent = _agent
+	baselogger.host = "please implement host_lookup()"
+	baselogger.version = "please implement version_lookup()"
+	baselogger.queue = _queue
+
+	//set options in priority order
+	baselogger.enabled = _enabled
+	//I believe comparing with an empty string should be the same as comparing with nil
+	if _url == "" {
+		baselogger.url = "urlByDefault, in the java version this is defined in UsageLoggers"
+		if baselogger.url == "" {
+			baselogger.enabled = false
+		} else {
+			baselogger.url = _url
+		}
+	}
+
+	//validate url when present
+	if baselogger.url != "" {
+		//I'm replacing the try catch statement with the error return
+		//but I'll implement that later, this line is a dummy
+		baselogger.url_parsed = baselogger.url + ";parsed"
+		//the rest is throwing errors
+	}
+	baselogger.enableable = (baselogger.url != "")
+
+	return baselogger
+}
+
+func (obj BaseLogger) enable() {
+	obj.enabled = true
+}
+
+func (obj BaseLogger) disable() {
+	obj.enabled = false
+}
+
+//Getters and Setters
+
+func (obj BaseLogger) getAgent() string         { return obj.agent }
+func (obj BaseLogger) getHost() string          { return obj.host }
+func (obj BaseLogger) getQueue() []string       { return obj.queue }
+func (obj BaseLogger) getSkipCompression() bool { return obj.skip_compression }
+func (obj BaseLogger) getSkipSubmission() bool  { return obj.skip_submission }
+func (obj BaseLogger) getUrl() string           { return obj.url }
+func (obj BaseLogger) getVersion() string       { return obj.version }
+func (obj BaseLogger) getEnableable() bool      { return obj.enableable }
+func (obj BaseLogger) getEnabled() bool         { return obj.enabled }
+func (obj BaseLogger) getSubmitFailues() int    { return obj.submit_failures }
```

Comment:
Here we will probably want to use `atomic.LoadInt64(&submitFailures)` Should look into sync/atomic package for atomic integers [](https://pkg.go.dev/sync/atomic#LoadInt64).